### PR TITLE
fix: joins on Timestamp columns

### LINKED ISSUE #187: [Datafusion] Support joins on TimestampMillisecond columns
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-11940

Joining DataFrames on a TimestampMillisecond column gives error:

```

'called `Result::unwrap()` on an `Err` value: Internal("Unsupported data type in hasher")

arrow/rust/datafusion/src/physical_plan/hash_join.rs:252:30

'

```

### PR DESCRIPTION
**Pull request by francis-du**
# Which issue does this PR close?

Closes #187

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by houqp — APPROVED** [review]

File: datafusion/src/physical_plan/hash_join.rs [hunk 1]
```diff
@@ -781,7 +782,7 @@ fn equal_rows(
             DataType::Float32 => equal_rows_elem!(Float32Array, l, r, left, right),
             DataType::Float64 => equal_rows_elem!(Float64Array, l, r, left, right),
             DataType::Timestamp(_, None) => {
-                equal_rows_elem!(Int64Array, l, r, left, right)
+                equal_rows_elem!(TimestampNanosecondArray, l, r, left, right)
```
Comment (Dandandan) [inline datafusion/src/physical_plan/hash_join.rs:785]: I think we also have to match on nanosecond unit in this case here (and on other units as well)

Comment (houqp) [reply datafusion/src/physical_plan/hash_join.rs:785]: oh yeah, good catch :+1:

Comment (francis-du) [reply datafusion/src/physical_plan/hash_join.rs:785]: I think other units need to first support the data types of these units in the Schema. And there is an issue about join on the `Time64(TimeUnit::Millisecond)` columns. I am analyzing this issue, and then I will open an issue and fix it.

image

File: datafusion/tests/sql.rs [hunk 2]
```diff
@@ -2124,6 +2124,49 @@ async fn cross_join_unbalanced() {
     );
 }
 
+#[tokio::test]
+async fn test_join_timestamp() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+
+    // register time table
```
Comment (alamb) [inline datafusion/tests/sql.rs:2131]: Also, the function `table_with_timestamps` might help writing tests for other timestamp units: 
 https://github.com/apache/arrow-datafusion/blob/master/datafusion/src/test/mod.rs#L194

**Review by alamb — APPROVED** [review]
Thanks for this contribution @francis-du  ! 

I think this PR could be merged as is (as it is strictly better than `master` in my opinion); If we don't add support for other `TimeUnits` then we can file a follow on ticket for thos case

**Review by Dandandan — APPROVED** [review]
Thanks, looks good @francis-du :+1:

**Comment by Dandandan** [comment]
I will merge it when all the tests pass.

**Comment by francis-du** [comment]
> I will merge it when all the tests pass.

Ok, thanks.

**Comment by francis-du** [comment]
Is this issue #959  related to this?

**Comment by Dandandan** [comment]
🎉

**Comment by alamb** [comment]
Thanks again @francis-du !

**Pull request merged**